### ISSUE TITLE
[Yul] CommonSubexpressionEliminator should take pure-ness of user-defined functions into account

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#7078]
Not only but especially with the introduction of the web assembly glue code, expressions often include user-defined functions. Think of endianness swapping code, for example. Currently, the CSE only considers built-in functions that are pure to not break this identity. We should extend pure-ness to user-defined functions such that CSE/Simplifier will optimize as follows (even without inlining):

```
function add_one(x) -> y { y := add(x, 1) }
let z := add_one(calldataload(0))
let t := add_one(calldataload(0))
let x := sub(z, t)
```
->

```
function add_one(x) -> y { y := add(x, 1) }
let x := 0
```

**Comment by chriseth** [comment]
The first step here would probably be to build a call graph between user-defined functions. Then determine the "basic" properties of user-defined functions depending on the built-in functions they use in their code. Finally, propagate these properties backwards in this call graph.

**Comment by chriseth** [comment]
This has been implemented in #7251
